XiEditor/XiDocumentController.swift

```diff
@@ -105,83 +118,63 @@ class XiDocumentController: NSDocumentController {
+    override func openDocument(_ sender: Any?) {
+        let urls = urlsFromRunningOpenPanel();
+        for url in urls ?? [] {
+            openDocumentAsync(atUrl: url)
+        }
     }
 
-    // this is called when documents are restored on startup
-    override func makeDocument(for urlOrNil: URL?, withContentsOf contentsURL: URL, ofType typeName: String) throws -> NSDocument {
-        if urlOrNil == contentsURL {
-            return try self.makeDocument(withContentsOf: contentsURL, ofType: typeName)
+    /// Asks core to open the specified document. On success, we will receive
+    /// the `new_view` notification.
+    func openDocumentAsync(atUrl url: URL?) {
+        if let url = url, let existing = self.document(for: url) {
+            existing.showWindows()
+        } else {
+            let delegate = (NSApp.delegate as! AppDelegate)
+            let params = url == nil ? [:] : ["file_path": url!.path]
+            delegate.dispatcher?.coreConnection
+                .sendRpcAsync("new_view_async", params: params)
         }
-        fatalError("Xi does not currently support document duplication")
     }
 
-    func setupDocument(_ document: Document, forUrl url: URL?) {
-        // we nil out this field to opt out of having NSDocument check for changes on disk when saving;
-        // we (theoretically) do that check in xi-core.
-        document.fileModificationDate = nil
-        Events.NewView(path: url?.path).dispatchWithCallback(document.dispatcher!) { (response) in
-            DispatchQueue.main.sync {
+    func makeDocumentSync(atUrl url: URL?) throws -> NSDocument {
+        let newDocument = try makeUntitledDocument(ofType: defaultType!) as! Document
+        Events.NewView(path: url?.path).dispatchWithCallback(newDocument.dispatcher!) { (response) in
+            DispatchQueue.main.async {
                 switch response {
                 case .ok(let result):
                     let viewIdentifier = result as! String
-                    self.setIdentifier(viewIdentifier, forDocument: document)
-                    document.coreViewIdentifier = viewIdentifier
+                    self.lock.lock()
+                    self.openViews[viewIdentifier] = newDocument
+                    newDocument.coreViewIdentifier = viewIdentifier
+                    newDocument.fileURL = url
+                    newDocument.fileModificationDate = nil
+                    self.lock.unlock()
                 case .error(let error):
-                    document.close()
+                    newDocument.close()
                     (NSApplication.shared.delegate as! AppDelegate).alert(text: error.message)
                 }
             }
         }
+        return newDocument
+    }
+
+    // Called when we create a new view, or when a new view is created at startup.
+    override func openUntitledDocumentAndDisplay(_ displayDocument: Bool) throws -> NSDocument {
+        let document = try makeDocumentSync(atUrl: nil)
```

Comment:
When a new untitled view is opened (at launch), the view has no plugins available when it should have. You can reproduce this simply by opening a blank view and try to look at the plugins menu.

After digging into this, I realised this was caused by the lack of view identifier for this new document until the handling method for available_plugin has already been executed.

Reply:
Okay cool, I had noticed sometimes the plugin menu was empty but hadn't tracked down a cause. Will look into that. 👍